conda_recipe_manager/parser/recipe_parser_convert.py

```diff
@@ -0,0 +1,360 @@
+"""
+File:           recipe_parser_convert.py
+Description:    Provides a subclass of RecipeParser that performs the conversion of a v0 recipe to the new v1 recipe
+                format. This tooling was originally part of the base class, but was broken-out for easier/cleaner code
+                maintenance.
+"""
+from __future__ import annotations
+
+from typing import Final, cast
+
+from conda_recipe_manager.parser._node import Node
+from conda_recipe_manager.parser._traverse import traverse
+from conda_recipe_manager.parser._types import (
+    ROOT_NODE_VALUE,
+    TOP_LEVEL_KEY_SORT_ORDER,
+    V1_BUILD_SECTION_KEY_SORT_ORDER,
+    Regex,
+)
+from conda_recipe_manager.parser._utils import stack_path_to_str, str_to_stack_path
+from conda_recipe_manager.parser.recipe_parser import RecipeParser
+from conda_recipe_manager.parser.types import CURRENT_RECIPE_SCHEMA_FORMAT, MessageCategory, MessageTable
+from conda_recipe_manager.types import JsonPatchType, JsonType, SentinelType
+
+
+class RecipeParserConvert(RecipeParser):
+    """
+    Extension of the base RecipeParser class to enables upgrading recipes from the old to new format.
+    This was originally part of the RecipeParser class but was broken-out for easier maintenance.
```

Comment:
How does breaking this out like this offer easier maintenance?

As far as I can tell the maintenance issue here is caused by the high cyclomatic complexity of the `render_to_new_recipe_format` method. A good rule of thumb is to aim for a complexity of 15 or lower but, as currently written, this method has a complexity of 49, much too high.

Furthermore if it is this easy to break this into a new class (and since `self` is only referenced once in the method) it would seem to me that this deserves to live as a function and not nested within a larger object:

```python
def convert_to_new_recipe_format(old_recipe: RecipeParser) -> tuple[str, MessageTable]:
    new_recipe = RecipeParser(self.render())
    ...
```

Reply:
The subclass is mostly there to make the static analyzer happy with accessing underscore functions. I also have future plans to break this up into more functions under this new "namespace".

As for maintenance, the original parser class is huge. It used to be an even larger class file and was split-up into a few supporting files and classes.

Out of curiosity, how did you calculate that? Do you have a plugin? I also think some of this logic is unavoidable with the nature of the conversion work. There is a lot of "if this, then that" to perform.

Reply:
Let me go and clean up a bit more here. I think I can have the message tracking and a few other pieces of state tracked as part of the class member that will simplify _some_ of the ick.

Reply:
I have McCabe linters enabled in my IDE. Since you're using `pylint` you can use https://pypi.org/project/pylint-mccabe/.

Reply:
I just posted a few commits that clean-up the inner-function and state usage. If you're okay with it, I'd like to look into this `mccabe` integration after this PR and #29 get merged in. I think that might cause a bigger refactor